src/graph.rs

```diff
@@ -288,40 +372,179 @@ impl ModuleGraph {
     Ok(())
   }
 
+  /// Return a vector of references to module objects in the graph. Only modules
+  /// that were fully resolved are present, as "errors" are omitted. If you need
+  /// to know what errors are in the graph, use the `.errors()` method, or if
+  /// you just need to check if everything is "ok" with the graph, use the
+  /// `.valid()` method.
+  #[cfg(feature = "rust")]
+  pub fn modules(&self) -> Vec<&Module> {
+    self
+      .module_slots
+      .iter()
+      .filter_map(|(_, ms)| match ms {
+        ModuleSlot::Module(m) => Some(m),
+        _ => None,
+      })
+      .collect()
+  }
+
+  /// Returns a map of the fully resolved dependency graph of the module graph.
+  #[cfg(feature = "rust")]
+  pub fn resolution_map(
+    &self,
+  ) -> HashMap<ModuleSpecifier, HashMap<String, Resolved>> {
+    let mut map = HashMap::new();
+    for (referrer, module_slot) in &self.module_slots {
+      if let ModuleSlot::Module(module) = module_slot {
+        let deps = module
+          .dependencies
+          .iter()
+          .map(|(s, dep)| (s.clone(), dep.maybe_code.clone()))
+          .collect();
+        map.insert(referrer.clone(), deps);
+      }
+    }
+    map
+  }
+
   /// Resolve a specifier from the module graph following any possible redirects
   /// returning the "final" module.
   pub fn resolve(&self, specifier: &ModuleSpecifier) -> ModuleSpecifier {
     let mut redirected_specifier = specifier;
+    let mut seen = HashSet::new();
+    seen.insert(redirected_specifier);
     while let Some(specifier) = self.redirects.get(redirected_specifier) {
+      if !seen.insert(specifier) {
+        eprintln!("An infinite loop of redirections detected.\n  Original specifier: {}", specifier);
+        break;
+      }
       redirected_specifier = specifier;
+      if seen.len() > 5 {
+        eprintln!("An excessive number of redirections detected.\n  Original specifier: {}", specifier);
+        break;
+      }
```

Comment:
FYI in CLI we have redirect limit set to 10

Reply:
Not where I borrowed the code from: https://github.com/denoland/deno/blob/11acdf1ea8b2978dcacad37b386f45e43a1f2748/cli/module_graph.rs#L1694-L1697

Reply:
I see, I was referring to this one: https://github.com/denoland/deno/blob/11acdf1ea8b2978dcacad37b386f45e43a1f2748/cli/file_fetcher.rs#L513